src/cargo/ops/cargo_doc.rs

```diff
@@ -55,6 +63,60 @@ pub struct DocOptions {
 pub fn doc(ws: &Workspace<'_>, options: &DocOptions) -> CargoResult<()> {
     let compilation = ops::compile(ws, &options.compile_opts)?;
 
+    if ws.gctx().cli_unstable().rustdoc_mergeable_info {
+        let now = std::time::Instant::now();
+        for (kind, doc_merge_info) in compilation.doc_merge_info.iter() {
+            let target_name = match kind {
+                CompileKind::Host => "host",
+                CompileKind::Target(t) => t.short_name(),
+            };
+            let (parts_paths, out_dir) = match doc_merge_info {
+                DocMergeInfo::None => continue,
+                DocMergeInfo::Fresh => {
+                    ws.gctx().shell().verbose(|shell| {
+                        shell.status("Fresh", format_args!("doc-merge for {target_name}"))
+                    })?;
+                    continue;
+                }
+                DocMergeInfo::Merge {
+                    parts_paths,
+                    out_dir,
+                } => (parts_paths, out_dir),
+            };
+
+            let mut cmd = ProcessBuilder::new(ws.gctx().rustdoc()?);
+            if ws.gctx().extra_verbose() {
+                cmd.display_env_vars();
+            }
+            cmd.retry_with_argfile(true);
+            // TODO: should we apply RUSTDOCFLAGS here?
+            cmd.arg("-o")
+                .arg(out_dir)
+                .arg("-Zunstable-options")
+                .arg("--merge=finalize");
```

Comment:
We don't add `--emit=toolchain-shared-resources` here. See <https://github.com/rust-lang/cargo/pull/16167/files#r2570359518>.